### DISCUSSION TITLE
Disclosure of where and how much AI generated code was used

### DISCUSSION THREAD

**Discussion by SonicRainbow** [stoatchat/discussions#1022] [Feedback]
I understand it's extremely difficult to avoid using AI as a programmer nowadays, even if you're going out of your way not to. However I noticed that Claude is a contributor in the main github repo and that is the only place I saw it and no other mention. I'm sure I and many others looking to switch to Stoat would be more reassured if a more formal statement was issued and if the use of AI and how much code was AI generated front and center on the github readme.

---

Maintainer edit: Thread is devolving into unrelated arguments, locking now.

**Comment by insertish** [comment]
At this moment, Stoat does **NOT** use any GenAi.

---

**UPDATE:** I just reverted everything because it was basically nothing anyways.
- https://github.com/stoatchat/for-web/pull/617 (no changes since it was already removed)
- https://github.com/stoatchat/for-desktop/pull/42
- https://github.com/stoatchat/service-openproject-zammad-sync/pull/2

**UPDATE 2:** forgot to mention, but if you do spot anything, just feel free to comment and let me know

Okay since this is being linked to, let me be very clear:
- GenAi is not being used to "build" (like it's not a vibe coded app or whatever, is probably the best way to put it) Stoat, it is almost entirely human written code by me and other contributors over the course of several years, the small exception being listed below (and other codegen from tools like lingui, etc. obviously)
- I have played with it at a very small scale a few times with this project, but I don't see any reason to continue to use it for any core component

The total contribution from Claude is clearly and transparently labeled as follows:
- ~~for-web, single commit, configuration: https://github.com/stoatchat/for-web/commit/f75eb3a10dc5815eb81634d16b53dd1447a26ca2~~ (apparently I already removed this when we switched workflows to `mise`)
- for-desktop, single commit, configuration: https://github.com/stoatchat/for-desktop/commit/3eb9b8e84bf05debf9843b80c468911fd095f4a0
- auxiliary service, zammad sync, single commit, tiny code gen: https://github.com/stoatchat/service-openproject-zammad-sync/commit/46ba6d753a63efd28be491202965c0cd074a4b60

I am not aware of any other GenAi use at this time

This comment was edited, a few times, I am quite all over the place, maybe I have provided reasonably transparency about this matter by now

**Reply by SonicRainbow** [reply]
That seems reasonable to me, and thanks for hearing me out

**Reply by DeclanChidlow** [reply]
Also worth noting here that Stoat started as a project long before consumer generative-AI capable of writing decent code came onto the scene. Stoat entered public beta in 2021. ChatGPT launched in late 2022.

Stoat is not vibe-coded slop, because Stoat's foundation and much of its larger development predates the very concept of vibe-coding. It was only very recently that gen-AI got to a point where something on the scale of Stoat *might* be possible.

You are more than welcome to audit the mind boggling amount of code that has gone into Stoat if you need evidence of the project being the product of humans, not AI.

**Reply by pieartsy** [reply]
Thank you for reverting everything!!

**Reply by Fluffies-System** [reply]
hey uh, here is a pretty big one:
https://github.com/stoatchat/stoat.chat/blob/dc114f0163994ed2bce54b302dcb03e3a6ce52b4/src/content/blog/long-live-stoat.md

if you tell us that blog post isn't AI generated we absolutely won't believe you. it makes us question what else was written using AI that didn't have claude marked as a contributor. we agree with another commenter suggesting yall should have a policy banning usage of AI code

**Comment by dunn** [comment]
for the record, the use of LLM-generated code is an absolute dealbreaker for me using or recommending stoat

**Reply by insertish** [reply]
Do you have any fine-grained reasons? Is this about reliability or something else?
Every line of code going into the codebase receives scrutiny.
Although, to my knowledge, I don't think any LLM-generated code has made it into the actual source code of either the backend or frontend. On my side, I've pretty much just been using it to automate some non-critical infrastructure work.

**Reply by dunn** [reply]
beyond the massive social and environmental damage caused by the deployment of LLMs at scale, i don't think 'i'll check the output of the machine designed to produce accurate-seeming bullshit' is the defense you think it is

**Reply by insertish** [reply]
I don't disagree with you nor am trying to defend the use of LLMs on that front, I just wanted to know if there was anything that I could address or otherwise clarify.

Just thought I'd add a little more, personally I am hopeful that open models that are both efficient and capable became more prevalent. I think most of the issues surrounding LLMs are generally tied to models being controlled and operated by huge corporations with little oversight. Unfortunately that is currently the status-quo and I cannot deny the usefulness of these tools for menial tasks which would otherwise take up time that I can put into programming and building software.

**Reply by SonicRainbow** [reply]
I hold similar views on AI but I am also aware that it's hard to avoid and that it does have genuine uses. That is why I requested disclosure, so that people could decide whether it's okay or not on their own. Responsible disclosure would be better than not knowing at all. However it is hard to separate "responsible" AI LLM use from the absolute blight that is what dunn describes being pushed by mega corporations and the government.

Still, it _is_ status quo, and unless more people decide to be contributors more and more code is going to be at least "AI assisted" as fewer and fewer people do work, especially on open source projects that don't charge users money nor show ads.

**Reply by ianling** [reply]
LLMs are in no way "hard to avoid". Everybody was doing just fine without accelerating the destruction of the planet in this particular way up until about a year ago.

For sure, live in accordance with your own principles first and foremost, but let's all agree to at least be in reality about the ethics, effects, and necessity of LLMs and where the data they use comes from.

**Reply by SonicRainbow** [reply]
I suppose what I meant by that is unless you're very specifically going out of your way to not use AI or LLMs they will creep in and promise easier ways to do things especially if you don't have anyone else helping you out as a lot of open source thankless projects tend to be. You even have platforms like steam now allowing AI code. As a tool it's great for *assisting* although way too many people rely on them way too much.

That being said, yes I do feel like the ethics and effects you see with AI data centers, lobbying groups, and LLM brainrot I feel like make it hard to  justify even the good tool that is there because of the society we live in.

The least that projects could do in this landscape is tell us how they use AI and where and then we can decide for ourselves to use them or not.

For the record I absolutely despise LLMs and what they've done and are doing, but I felt it out of place to demand a project I don't contribute to to stop using it altogether and more realistic to ask them to be transparent about it.

**Reply by aw-tusk** [reply]
I'm not entirely sure what is meant by code that LLMs "excel" at. LLMs helping with code aren't really to a science to my knowledge; It's only whatever the LLM happens to "know" and decide to say. I do hope the LLM-generated code could be revised either way. 

I wont comment on the environmental or economic effect because my knowledge on that only extends to a Drew Gooden video because I try to minimize my involvement with the recent push of AI for the most part. LLMs have been more reliable than a decade ago, sure, but that's only from an uncountable amount of iterations. Even with careful use, it could not be consistent because it's taking from different sources with little moderation. Either way, probably needs to be rewritten to maintain a certain coding "style." That's how neural networks go, no matter how fine-tuned they may be. 

Coding isn't only about functionality. Experience with code is also important. Every line could be written with intent, even if it feels less significant. With how many caveats it has, I have to question the use of LLMs at all with a project as large as this.

**Reply by phanomgames** [reply]
you aren't minimizing your use of LLM products. steam, every aaa game from now on, erp software that runs your taxes, literally everything you interact with going forward will now be at least partially assisted with llms. And guess what, discord and teamspeak are too, so there is no alternate path unless you're going to fire up a ts3 server

**Reply by VaelophisNyx** [reply]
> you aren't minimizing your use of LLM products. steam, every aaa game from now on, erp software that runs your taxes, literally everything you interact with going forward will now be at least partially assisted with llms. And guess what, discord and teamspeak are too, so there is no alternate path unless you're going to fire up a ts3 server

But why the defensive lies though

**Reply by damejigglebones** [reply]
I do not understand the people claiming that LLMs are "extremely difficult to avoid." They aren't coding tools. They are predictive text. Yk how your phone's autocorrect is annoying and gets things wrong? Why are you trusting _that_ with your _code_?

**Reply by Quadstrike** [reply]
i think this thread is getting a creep of people who haven't used LLM assistance tools for coding, so it's slowly devolving.
to clarify for those who don't know: LLM assistance tools that are NOT generating/vibe coding are time savers for having to search through stack overflow or something for hours just to figure out you forgot to close a bracket, or are referencing a function that hasn't been defined. if you feel that getting that sort of help is outside of your realm of acceptability, that's fine, but understand this is the coding equivalent to when Wikipedia first started, vs going to the library and reading through dated publications that may have misinformation. a coder can try an LLM's bad solutions far faster than sifting through searches remotely related to their problem, that's assuming their are ANY posts applying to their use case, and waiting on people to possibly respond on a forum or stack overflow can take days/weeks/months depending.
LLM can be self hosted and simple, so that can answer some of the other concerns

hope this helps some confusion for anyone who doesn't know

**Reply by Darxoon** [reply]
@Quadstrike it's ironic that you claim to be so much more knowledged about LLMs and then say that they help you when you forgot to close a bracket or are referencing a function that doesn't exist. in those cases, your editor/LSP will already yell at you way quicker than you can ask an AI for help :skull:

**Reply by phanomgames** [reply]
the best part is, everyone in this thread is using github when 30%+ of the new code written by microsoft products (including github) is ai generated. Ah, but you guys don't want to actually minimize your exposure to ai generated code, you just want to virtue signal. Feel free to go ahead and stand up your own git/fork stoat

But wait! linus torvalds is a big fan of all out vibe coding! So you should back up and make another alternative to git, right? wouldn't want to support people who use AI either.

Childish and pathetic. LLM's are a tool, a really good time saver.

**Reply by Quadstrike** [reply]
> @Quadstrike it's ironic that you claim to be so much more knowledged about LLMs and then say that they help you when you forgot to close a bracket or are referencing a function that doesn't exist. in those cases, your editor/LSP will already yell at you way quicker than you can ask an AI for help 💀

i was grossly simplifying for those of us who don't have any frame of reference, i didn't claim to be more knowledgeable. being irrationally aggressive and nitpicky on someone trying to make some sort of clarification won't help anyone, you're welcome to clarify it if you're so inclined, but if you just want to give the impression that vibecoding whole programs is no different than using an LLM to debug, you're simply mistaken and actively misinforming

**Reply by rstat1** [reply]
> i think this thread is getting a creep of people who haven't used LLM assistance tools for coding, so it's slowly devolving. to clarify for those who don't know: LLM assistance tools that are NOT generating/vibe coding are time savers for having to search through stack overflow or something for hours just to figure out you forgot to close a bracket, or are referencing a function that hasn't been defined. if you feel that getting that sort of help is outside of your realm of acceptability, that's fine, but understand this is the coding equivalent to when Wikipedia first started, vs going to the library and reading through dated publications that may have misinformation. a coder can try an LLM's bad solutions far faster than sifting through searches remotely related to their problem, that's assuming their are ANY posts applying to their use case, and waiting on people to possibly respond on a forum or stack overflow can take days/weeks/months depending. LLM can be self hosted and simple, so that can answer some of the other concerns
> 
> hope this helps some confusion for anyone who doesn't know

Why is it there's always at least one person in these threads that always pulls the "oh you must not have used it" card when ever someone criticizes these slop machines? No that's almost never the case.

>  coder can try an LLM's bad solutions far faster than sifting through searches remotely related to their problem

This is just completely untrue. For someone who actually knows that they're doing (aka real programmers, not pretenders under the guise of "vibe coding") its at best the same of amount of time, and at worse some multiple of the amount of time you'd have spent either figuring it out yourself or searching for a solution. And if you actually figure it out yourself then you actually learned something.

I will freely concede these slop machines are great at generating a good starting point for documentation for a project

**Comment by StevenStavrakis** [comment]
Please do NOT make a statement on this. If someone is unwilling to use your platform because they don't know how much AI was involved, then they probably weren't going to use your platform anyway. Simply make good software.

Making statements on how much AI you use, plan to use, will use, commit to using, etc. can only do harm. The obvious feedback that matters is: don't rely explicitly on AI such that you guys don't know what the software is doing. Not for any moral reasons, but because you need a deep understanding of your codebase to keep making great improvements. But you already know that.

Keeping my eye on this project in light of the new Discord ID verification policies.

**Reply by VaelophisNyx** [reply]
Using 0% slop code is the only way. Hard stop, no nuance or wiggle room.

**Reply by python357-1** [reply]
The only thing Gen AI has proven thus far is that it can generate plenty of security issues. If I'm going to ask my friend group to switch to this, I need to know they're going to do better than that.

**Reply by akhra** [reply]
> If someone is unwilling to use your platform because they don't know how much AI was involved, then they probably weren't going to use your platform anyway.

I'm here, in this specific thread, to help decide whether to endorse this platform for multiple communities. which the developer response(/actions) above made *more* likely.

**Comment by marrus-sh** [comment]
I have a Jewish friend who compares avoiding A·I use to keeping kosher. For Catholics, the Pope has recently urged that we foreground human faces, voices, and thoughts over those generated by machines. While this is not a full‐throated condemnation of A·I, it is not hard to imagine a person wanting to ensure that all of the thoughts in the programs they run were intentionally created by humans or as the direct result of human processes. When an L·L·M is used, this is simply impossible to maintain.

I ask that people understand that there may be religious objections to programs made with A·I and that this is no different from having religious objections to food prepared a certain way or with certain ingredients. As we approach the season of Lent, I ask that people respect these religious choices and label their products so that people can make informed decisions about the software they use. And understand that there will always be a market for programs which are free of A·I, just as there will always be a market for food that is meat‐free or kosher or halal. Trying to get a corner on that market is not a bad idea for an open‐source project seeking to differentiate itself from competitors.

**Comment by Azure-Lazuline** [comment]
regardless of the many morality-based arguments you can make against LLM usage, it's worth considering that there are many people with _incredibly_ strong anti-LLM views, and those people tend to be minorities or other vulnerable groups. if it's only 0.01% of the codebase anyway, you should heavily consider putting in a little extra work to make that 0% and ensure that it stays that way: both from a business perspective to gain more users, and also just because it's a nice thing to do for people who are already struggling immensely with a world that's becoming more and more LLM-centric. thank you for reading and considering this, if you do.

**Reply by insertish** [reply]
I'm happy to revert the commits since it's not all that much, once GitHub starts working anyways, since it's not all that much.

**Reply by VaelophisNyx** [reply]
> I'm happy to revert the commits since it's not all that much, once GitHub starts working anyways, since it's not all that much.

I would heavily recommend you do and discontinue use of it at all and ban use of it in the future for the project at large. LLM slop code is the death of projects like this.

**Reply by Machinatory** [reply]
hopefully you can also hold together. good luck out here, things are gonna move fast.

**Reply by insertish** [reply]
It's gone.

**Reply by Delta-Kruger** [reply]
Alas, the sheer fact that LLM slop-code has touched it at all is bound to be a black stain on its record, but im glad that one is still willing to immediately cull such things

**Reply by akhra** [reply]
purity tests serve nobody. people make mistakes. it's the follow-up that matters.

**Comment by includenull** [comment]
Even if contributors were using LLMs in their IDEs or using cli tools you wouldn't nessesarily know. It doesn't matter where the code comes from as long as it is tested and audited.

**Comment by pieartsy** [comment]
I think it might be worth closing this now that the genai stuff has been reverted and there's no need for further discussion.

**Reply by akhra** [reply]
was on the fence about this suggestion when I first read it, but now finding myself fighting to resist an urge to dunk on other commenters... yeah I think the good work has been done and the thread is quickly sliding into social media cognitohazard territory. I'm out, cheers to the team!

**Comment by JmeJuniper** [comment]
I appreciate the response, but I think the root of this discussion hasn't been fulfilled (unless I'm missing something)—it would be great to have a dedicated policy on the usage of generative AI tools in Stoat. This would serve the purpose of assuring people who may worry it does have LLM code that that is not the case, ensure that new contributors are aware that they should not do so, and (of course, IANAL, TINLA) would likely help you as the status of LLM-generated code legally is dubious to say the least.

**Comment by MattyKatos** [comment]
I do not understand why people are acting like audited AI generated code is the root of all evil. A small dev team does not have 100s of entry level coders that can write tedious and simple code for them. Using AI to auto generate code is acceptable and should not be treated the way it's being treated. It's not like you using AI on this project is taking jobs away from people. If they want you to not use gen AI, they had better be financially supporting the team significantly. These people are using Windows that is now running off a significant amount of AI code, they're using google, with integrated AI, they use the assistants on their smart phone that have all moved to AI. This straw man bullshit has to stop. Let the fucking developers develop. If you see issues in the open source code generated by said AI, that is a TOTALLY different story. That should be addressed and the way that the devs are using AI should be monitored to ensure they aren't vibe coding.

**Reply by rstat1** [reply]
>A small dev team does not have 100s of entry level coders that can write tedious and simple code for them.

100s of entry level coders? Really? People have been writing "simple and tedious code" since the dawn of computing, and never needed any of the slop machines to do it.

**Reply by JmeJuniper** [reply]
Hihi! Bold claims you're making.
* I'm not running Windows
* I use google only when I have to, and even then never the AI features
* I don't even have a smartphone 😂

So, with whatever weird adhoms you're trying to preemptively throw around out of the way, let's address what actual points you've made.

> A small dev team does not have 100s of entry level coders that can write tedious and simple code for them.

As has been mentioned, many, many small teams have functioned perfectly fine without GenAI for quite a while. So, here in real programming, we have these cool things called "libraries". The gist of them is that if everyone keeps writing the same boilerplate code with very few differences, that could instead be put into a little chunk of code that your program imports. Then, the program relies on the library to handle the tedious and simple code, while it can develop the high level application! Easy!

> It's not like you using AI on this project is taking jobs away from people.

It's a good thing nobody has made the argument that Stoat specifically using GenAI will take jobs away from people.


Now, you've made a lot of claims about audited AI usage. Surely, if you're advocating for it, there must be some good reason? Clearly, it must massively speed up developer productivity, since they no longer have to write the equivalent of "100s of entry level coders" of "tedious and simple code", right? Well, here in the real world, usage of AI actually _slows down_ developers<sup>[1]</sup>, despite their thinking it would speed them up. This should make intuitive sense as well, given that anyone who has reviewed others' code knows that understanding code is more difficult if you haven't written it. Thus, if you want developers to understand the code that they are submitting (something that should be absolutely MANDATORY for a project), having GenAI create code is going to take longer than them just. writing the code themselves.

Look, sorry if this whole thing comes off as patronizing, but as someone who has _actually_ been working with AI for... about a decade, now, it upsets me when people spout talking points they don't actually know anything about.

1: https://arxiv.org/pdf/2507.09089

**Reply by phanomgames** [reply]
Lol, you use linux? Hate to break it to you, but you're gonna have to find a new os then. Maybe bsd doesn't have ai code yet. 
<img width="679" height="314" alt="image" src="https://github.com/user-attachments/assets/6f93f04f-f216-4b5f-ba5e-81c34d59ac78" />

**Reply by ctrldemi** [reply]
hey nerd, you know that screenshot is from one of Linus' side projects and this won't ever affect the linux repo, right?
